arch/x86_64/mcount-dynamic.c

```diff
@@ -145,21 +145,16 @@ static void read_patchable_loc(struct mcount_dynamic_info *mdi, struct uftrace_e
 			       struct uftrace_elf_iter *iter, unsigned long offset)
 {
 	typeof(iter->shdr) *shdr = &iter->shdr;
+	unsigned i;
+	unsigned long *patchable_loc;
 
 	mdi->nr_patch_target = shdr->sh_size / sizeof(long);
 	mdi->patch_target = xmalloc(shdr->sh_size);
+	patchable_loc = mdi->patch_target;
 
-	elf_get_secdata(elf, iter);
-	elf_read_secdata(elf, iter, 0, mdi->patch_target, shdr->sh_size);
-
-	/* symbol has relative address, fix it to match each other */
-	if (elf->ehdr.e_type == ET_EXEC) {
-		unsigned long *patchable_loc = mdi->patch_target;
-		unsigned i;
-
-		for (i = 0; i < mdi->nr_patch_target; i++) {
-			patchable_loc[i] -= offset;
-		}
+	for (i = 0; i < mdi->nr_patch_target; i++) {
+		unsigned long *entry = (unsigned long *)(offset + shdr->sh_addr) + i;
+		patchable_loc[i] = *entry - offset;
 	}
```

Comment:
I tested it and it works fine in default build, which is `ET_DYN` as follows.
```
$ gcc -o t-abc.dyn -fpatchable-function-entry=5 tests/s-abc.c 

$ readelf -h t-abc.dyn | grep Type
  Type:                              DYN (Position-Independent Executable file)

$ uftrace -P. --debug-domain dynamic:4 t-abc.dyn 
dynamic: dynamic patch type: t-abc.dyn: 5 (fpatchable)
dynamic: update 0x56351b77616d for 'a' function dynamically to call __fentry__
dynamic: update 0x56351b776184 for 'b' function dynamically to call __fentry__
dynamic: update 0x56351b77619b for 'c' function dynamically to call __fentry__
dynamic: update 0x56351b7761d3 for 'main' function dynamically to call __fentry__
dynamic: patched all (4) functions in 't-abc.dyn'
# DURATION     TID     FUNCTION
            [ 22119] | main() {
            [ 22119] |   a() {
            [ 22119] |     b() {
            [ 22119] |       c() {
   0.886 us [ 22119] |         getpid();
   2.081 us [ 22119] |       } /* c */
   2.306 us [ 22119] |     } /* b */
   2.529 us [ 22119] |   } /* a */
   3.016 us [ 22119] | } /* main */
```
But having `-no-pie` option makes the executable type `ET_EXEC` then it gets segfaulted as follows.
```
$ gcc -o t-abc.exec -no-pie -fpatchable-function-entry=5 tests/s-abc.c 

$ readelf -h t-abc.exec | grep Type
  Type:                              EXEC (Executable file)

$ uftrace -P. --debug-domain dynamic:4 t-abc.exec 
WARN: child terminated by signal: 11: Segmentation fault
WARN: cannot open record data: /tmp/uftrace-live-VpBStd: No data available
```
Please check if `shdr->sh_addr` is correct in this case.

Reply:
> But having `-no-pie` option makes the executable type `ET_EXEC` then it gets segfaulted as follows.
...
> Please check if `shdr->sh_addr` is correct in this case.

Yeah, looks like for ET_EXECs the addresses in section are absolute so we do not need the `offset` addend...

Reply:
Thanks for the update! I've just tested and it works fine on both x86_64 and aarch64.

Reply:
@namhyung Please have a final look.

Reply:
@yugr It'd be better if you could rebase it targeting to the current master.

Reply:
Done!